src/librustc_typeck/check/wfcheck.rs

```diff
@@ -638,24 +648,37 @@ fn report_bivariance<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     err.emit();
 }
 
-fn reject_shadowing_type_parameters(tcx: TyCtxt, def_id: DefId) {
+fn reject_shadowing_parameters(tcx: TyCtxt, def_id: DefId) {
     let generics = tcx.generics_of(def_id);
     let parent = tcx.generics_of(generics.parent.unwrap());
-    let impl_params: FxHashMap<_, _> = parent.types
-                                       .iter()
-                                       .map(|tp| (tp.name, tp.def_id))
-                                       .collect();
-
-    for method_param in &generics.types {
-        if impl_params.contains_key(&method_param.name) {
+    let impl_params: FxHashMap<_, _> =
+        parent.params.iter()
+                     .flat_map(|param| {
+                         match param {
+                             GenericParamDef::Lifetime(_) => None,
+                             GenericParamDef::Type(ty) => Some((ty.name, ty.def_id)),
+                         }
+                     })
+                     .collect();
+
+    for method_param in generics.params.iter() {
+        // Shadowing is currently permitted with lifetimes.
```

Comment:
It's not, it's just checked elsewhere (because of course it is, *sigh*).

Reply:
Ah, I see it. I was getting shadowing and unconstrained parameters mixed up.